### PR TITLE
V7 to V8 Package Developer Port Guide - Working Document

### PR DESCRIPTION
**Pull request by warrenbuckley**
Hi Docs PR Team,
I have created an initial document of notes & pointers on how to port an Umbraco V7 package to V8. I am sure the Umbraco Package Developer community will have more suggestions to add.

I would love it if you could review it for grammar & anything super obvious that needs cleaning up.
@Jeavon @marcemarc & @dampee - I would love your technical input to see if there is anything that does not make sense or I may have missed or could be added to help improve the porting process.

### CODE REVIEW

File: Tutorials/index.md [hunk 1]
```diff
@@ -27,3 +27,7 @@ Learn how you can use your Google user to log into the Umbraco backoffice.
 ### [Starter kit](Starter-kit/index.md)
 
 Information about the basic Umbraco starter kit.
+
+### [Porting Packages from Version 7 to Version 8](Porting-Packages-V8/index.md)
+
+A working document, that gives tips & pointers on how to port an Umbraco V7 package to work with the upcomign Umbraco V8 release.
```
Comment (jmayntzhusen) [inline Tutorials/index.md]:
```suggestion
A working document, that gives tips & pointers on how to port an Umbraco V7 package to work with the upcoming Umbraco V8 release.
```

File: Tutorials/Porting-Packages-V8/index.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
+# Porting Umbraco V7 packages to work with V8
+This is not a definitive guide for all scenarios and more of a working document and will evolve over time, especially whilst V8.0.0 is development and things in this guide may change.
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]:
```suggestion
This is not a definitive guide for all scenarios and more of a working document and will evolve over time, especially whilst V8.0.0 is in development and things in this guide may change.
```

File: Tutorials/Porting-Packages-V8/index.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+Well then this guide, should give you a few pointers to get started with the process and is mostly geared towards solutions that use C#
+
+## Upgrade .NET Framework
+Upgrade your Visual Studio C# project to use and compile against the .NET Framework 4.7.2. This is because the minimum to run version 8 of Umbraco requires this specific version of the .NET Framework
+
+## Add Nuget.config file to your solution
+As Umbraco V8 is still in development and not been published to the official Nuget.org feed, then to allow you to easily add references to the Umbraco DLLs in your project.
+
+Add the following `nuget.config` file to the root of your project, for Visual Studio to discover it
+
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]: Our has syntax highlighting now, you should wrap this code block in 
````
```xml
```
````

File: Tutorials/Porting-Packages-V8/index.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                }
+            }
+        }
+    }
+
+
+## Update package.xml for Umbraco ZIP packages
+Currently the Umbraco packager in the backoffice is still undergoing changes (at the time of writing) and are unable to use the friendly packager UI in Umbraco.
+
+As version 7 and version 8 of Umbraco has breaking changes in its APIs and code, its recommended to ship a different version of your package to support V8 only. So update the package.xml in your Umbraco package zip to have
```
Comment (Shazwazza) [inline Tutorials/Porting-Packages-V8/index.md]: @warrenbuckley this sentence doesn't seem to be completed

File: Tutorials/index.md (hunk 1 above)
Comment (Shazwazza) [inline Tutorials/index.md]: up coming instead of upcomign

File: Tutorials/Porting-Packages-V8/index.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    <packageSources>
+      <add key="UmbracoCoreMyGet" value="https://www.myget.org/F/umbracocore/api/v3/index.json" />
+      <add key="ExamineAppVeyor" value="https://ci.appveyor.com/nuget/examine-f73l6qv0oqfh/" />
+    </packageSources>
+  </configuration>
+
+
+## Add references via Nuget
+You will need to add `UmbracoCms.Core` via Nuget from the `UmbracoCoreMyGet` feed with the pre-release flag enabled.
+
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]: I guess package devs would know the command for this, but maybe we should add it here just to be nice? 🤔
`Install-Package UmbracoCms.Core -pre -source https://www.myget.org/F/umbracocore/api/v3/index.json`

File: Tutorials/Porting-Packages-V8/index.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+Optionally add reference to `UmbracoCms.Web` if your code uses anything such as `UmbracoApiController` `SurfaceController` then you will need to add the Nuget package reference of `UmbracoCms.Web` which in turn has a reference to `UmbracoCms.Core`
+
+## Try & build the solution
+OK now the references have been updated, try and build the solution and then
+
+## Fix up compile errors
+As long as your Umbraco package was not using ancient APIs and using the new APIs & Services in Umbraco V7 then most of this step will be trying to determine what the new V8 namespace equivelant is.
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]:
```suggestion
As long as your Umbraco package was not using ancient APIs and using the new APIs & Services in Umbraco V7 then most of this step will be trying to determine what the new V8 namespace equivalent is.
```

File: Tutorials/Porting-Packages-V8/index.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+### Handy tip
+Clone down the Umbraco CMS source code and checkout the `temp8` branch and use this as a way to try and find that same method that the compiler is complaining about, in it's new namespace.
+
+## Bye bye ApplicationEventHandler, Hello Composer & Components
+The common thing that most package developers and Umbraco developers use in their projects is an `ApplicationEventHandler` to do some logic at the startup of Umbraco or to register event listeners based on Umbraco events such as ContentService.Saved etc.
+
+In Umbraco version 8 this has been removed and replaced with Composer & Components.
+
+### Old ApplicationEventHandler way
+
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]: Our has syntax highlighting now, you should wrap this code block in 
````
```csharp
```
````

File: Tutorials/Porting-Packages-V8/index.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+
+
+### New Composer & Component way
+
```
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]: Our has syntax highlighting now, you should wrap this code block in 
````
```csharp
```
````

File: Tutorials/Porting-Packages-V8/index.md (hunk 4 above)
Comment (jmayntzhusen) [inline Tutorials/Porting-Packages-V8/index.md]: Seems like half of the sentence is missing here? 🤔

File: Tutorials/Porting-Packages-V8/index.md (hunk 5 above)
Comment (warrenbuckley) [reply Tutorials/Porting-Packages-V8/index.md]: Well I still use the UI in Visual Studio to add/manage my Nuget package references. But yes a valid point will add the command :)

File: Tutorials/Porting-Packages-V8/index.md (hunk 4 above)
Comment (warrenbuckley) [reply Tutorials/Porting-Packages-V8/index.md]: @jmayntzhusen & @Shazwazza I have updated the missing copy & example XML from this section.

**Comment by jmayntzhusen** [comment]
Only thing missing now is YAML meta data, I will add it later today when time permits and then merge 😄👍

**Comment by marcemarc** [comment]
@warrenbuckley @Shazwazza I have a naive question regarding the new approach to wire up changes at startup with an IUserComposer and Composition and that is - is this common parlance in this kind of approach? if not, and these are new terms, can we change them?

Essentially 'Composition' as a word is very synonymous with another part of Umbraco, namely composing document types - whilst this is a very different part of Umbraco, to this code, (and people will just cut and paste whatever is here) my observation is when people are searching for help on 'umbraco compositions' they have a bit of noise to wade through - so if we can avoid having clashing terms within the Umbraco domain language system then I think we should. (hence the naivety as not sure if this is just what these thinks are called, and it would be more confusing to call them something else). And also technically a composition is named a 'Document Type without Template' anyway, I'm not sure officially compositions exist as a term, though we talk about them all the time.

Finally if this isn't a common parlance thing, then in my understanding of what we're doing here, we are not composing, we are not 'creating something' we are wiring stuff up, pulling into play custom implementations - so if you want to go musical with your analogy then I would suggest calling this the **Conductor** class, as really at this stage you are **Conducting** waving your coding baton and deciding which parts of your Orchestra will take part and play which parts of the Umbraco Score... subtly different but maybe easier to grok?

**Pull request merged**